network/src/gossip/attestation.rs

```diff
@@ -0,0 +1,239 @@
+#[derive(Default)]
+pub(super) struct Knowledge {
+	candidates: HashSet<Hash>,
+}
+
+impl Knowledge {
+	// whether the peer is aware of a candidate with given hash.
+	fn is_aware_of(&self, candidate_hash: &Hash) -> bool {
+		self.candidates.contains(candidate_hash)
+	}
+
+	// note that the peer is aware of a candidate with given hash.
+	fn note_aware(&mut self, candidate_hash: Hash) {
+		self.candidates.insert(candidate_hash);
+	}
+}
+
+#[derive(Default)]
+pub(super) struct PeerData {
+	live: HashMap<Hash, Knowledge>,
+}
+
+impl PeerData {
+	/// Update leaves, returning a list of which leaves are new.
+	pub(super) fn update_leaves(&mut self, leaves: &LeavesVec) -> LeavesVec {
+		let mut new = LeavesVec::new();
+		self.live.retain(|k, _| leaves.contains(k));
+		for &leaf in leaves {
+			self.live.entry(leaf).or_insert_with(|| {
+				new.push(leaf);
+				Default::default()
+			});
+		}
+
+		new
+	}
+
+	#[cfg(test)]
+	pub(super) fn note_aware_in_session(&mut self, relay_chain_leaf: &Hash, candidate_hash: Hash) {
+		if let Some(knowledge) = self.live.get_mut(relay_chain_leaf) {
+			knowledge.note_aware(candidate_hash);
+		}
+	}
+
+	pub(super) fn knowledge_at_mut(&mut self, parent_hash: &Hash) -> Option<&mut Knowledge> {
+		self.live.get_mut(parent_hash)
+	}
+
+	/// Get an iterator over all live leaves of this peer.
+	pub(super) fn leaves(&self) -> impl Iterator<Item = &Hash> {
+		self.live.keys()
+	}
+}
+
+/// An impartial view of what topics and data are valid based on attestation session data.
+pub(super) struct View {
+	live_sessions: Vec<(Hash, SessionView)>,
+	topics: HashMap<Hash, Hash>, // maps topic hashes to block hashes.
+}
+
+impl Default for View {
+	fn default() -> Self {
+		View {
+			live_sessions: Vec::with_capacity(MAX_CHAIN_HEADS),
+			topics: Default::default(),
+		}
+	}
+}
+
+impl View {
+	fn session_view(&self, relay_chain_leaf: &Hash) -> Option<&SessionView> {
+		self.live_sessions.iter()
+			.find_map(|&(ref h, ref sesh)| if h == relay_chain_leaf { Some(sesh) } else { None } )
+	}
+
+	fn session_view_mut(&mut self, relay_chain_leaf: &Hash) -> Option<&mut SessionView> {
+		self.live_sessions.iter_mut()
+			.find_map(|&mut (ref h, ref mut sesh)| if h == relay_chain_leaf { Some(sesh) } else { None } )
+	}
+
+	/// Get our leaves-set. Guaranteed to have length <= MAX_CHAIN_HEADS.
+	pub(super) fn neighbor_info<'a>(&'a self) -> impl Iterator<Item=Hash> + 'a + Clone {
+		self.live_sessions.iter().take(MAX_CHAIN_HEADS).map(|(p, _)| p.clone())
+	}
+
+	/// Note a new session.
+	pub(super) fn add_session(&mut self, relay_chain_leaf: Hash, validation_data: MessageValidationData) {
+		self.live_sessions.push((
+			relay_chain_leaf,
+			SessionView {
+				validation_data,
+				knowledge: Default::default(),
+			},
+		));
+		self.topics.insert(attestation_topic(relay_chain_leaf), relay_chain_leaf);
+	}
+
+	/// Prune old sessions that fail the leaf predicate.
+	pub(super) fn prune_old_sessions<F: Fn(&Hash) -> bool>(&mut self, is_leaf: F) {
```

Comment:
again, not super clear what this is meant to do. "sessions" clashes with runtime terminology and `is_leaf`'s meaning is unclear (should it be renamed to `keep_leaf`?).